File: ext/nokogiri/html_sax_parser.c
#include <html_sax_parser.h>

/*
 * call-seq:
 *  native_parse_file(data, encoding)
 *
 * Parse +data+ with +encoding+
 */
static VALUE native_parse_file(VALUE self, VALUE data, VALUE encoding)
{
  xmlSAXHandlerPtr handler;
  htmlDocPtr hdoc ;
  Data_Get_Struct(self, xmlSAXHandler, handler);
  hdoc = htmlSAXParseFile( StringValuePtr(data),
                           (const char *)StringValuePtr(encoding),
                           (htmlSAXHandlerPtr)handler,
                           (void *)self );
  xmlFreeDoc(hdoc);
  return data;
}

/*
 * call-seq:
 *  native_parse_memory(data, encoding)
 *
 * Parse +data+ with +encoding+
 */
static VALUE native_parse_memory(VALUE self, VALUE data, VALUE encoding)
{
  xmlSAXHandlerPtr handler;
  htmlDocPtr hdoc ;
  Data_Get_Struct(self, xmlSAXHandler, handler);
  hdoc = htmlSAXParseDoc(  (xmlChar *)StringValuePtr(data),
                           (const char *)StringValuePtr(encoding),
                           (htmlSAXHandlerPtr)handler,
                           (void *)self );
  xmlFreeDoc(hdoc);
  return data;
}

VALUE cNokogiriHtmlSaxParser ;
void init_html_sax_parser()
{
  VALUE nokogiri  = rb_define_module("Nokogiri");
  VALUE html      = rb_define_module_under(nokogiri, "HTML");
  VALUE sax       = rb_define_module_under(html, "SAX");
  /*
   * Nokogiri::HTML::SAX::Parser is used for parsing HTML with SAX
   * callbacks.
   */
  VALUE klass     = rb_define_class_under(sax, "Parser", cNokogiriXmlSaxParser);

  cNokogiriHtmlSaxParser = klass;

  rb_define_private_method(klass, "native_parse_memory", native_parse_memory, 2);
  rb_define_private_method(klass, "native_parse_file", native_parse_file, 2);
}

File: ext/java/nokogiri/HtmlSaxParser.java
package nokogiri;

import java.io.IOException;
import java.io.StringReader;
import java.util.logging.Level;
import java.util.logging.Logger;
import nokogiri.internals.NokogiriHandler;
import org.cyberneko.html.parsers.SAXParser;
import org.jruby.Ruby;
import org.jruby.RubyClass;
import org.jruby.RubyObject;
import org.jruby.anno.JRubyMethod;
import org.jruby.exceptions.RaiseException;
import org.jruby.runtime.ThreadContext;
import org.jruby.runtime.Visibility;
import org.jruby.runtime.builtin.IRubyObject;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import org.xml.sax.ext.DefaultHandler2;

public class HtmlSaxParser extends XmlSaxParser {
    private DefaultHandler2 handler;
    private SAXParser parser;

    public HtmlSaxParser(Ruby ruby, RubyClass rubyClass) {
        super(ruby, rubyClass);

        this.handler = new NokogiriHandler(ruby, this);

        
            this.parser = new SAXParser();

            this.parser.setContentHandler(this.handler);
            this.parser.setErrorHandler(this.handler);
            try{
                this.parser.setProperty("http://xml.org/sax/properties/lexical-handler", handler);
                this.parser.setProperty("http://cyberneko.org/html/properties/names/elems", "match");
                this.parser.setProperty("http://cyberneko.org/html/properties/names/attrs", "no-change");
            } catch(Exception ex) {
                System.out.println("Problem while creating HTML SAX Parser: "+ex.toString());
            }
        
    }

    @JRubyMethod(visibility = Visibility.PRIVATE)
    public IRubyObject native_parse_memory(ThreadContext context, IRubyObject self, IRubyObject data, IRubyObject encoding) {
        String input = data.convertToString().asJavaString();
        try {
            this.parser.parse(new InputSource(new StringReader(input)));
        } catch (SAXException se) {
            throw RaiseException.createNativeRaiseException(context.getRuntime(), se);
        } catch (IOException ioe) {
            throw context.getRuntime().newIOErrorFromException(ioe);
        }
        return data;
    }

    @JRubyMethod(visibility = Visibility.PRIVATE)
    public IRubyObject native_parse_file(ThreadContext context, IRubyObject self, IRubyObject data, IRubyObject encoding) {
        String file = data.convertToString().asJavaString();
        try {
            this.parser.parse(file);
        } catch (SAXException se) {
            throw RaiseException.createNativeRaiseException(context.getRuntime(), se);
        } catch (IOException ioe) {
            throw context.getRuntime().newIOErrorFromException(ioe);
        }
        return data;
    }
}
